concepts/storage-class-specifiers/about.md

```diff
@@ -0,0 +1,184 @@
+// as a variable with static storage, it is initialized to zero
+int val;
+```
+
+If both `extern` keywords were removed the program might print something like
+
+```
+val is 22038
+val is 42
+```
+
+Such an output demonstrates that each declaration of `val` without `extern` is a defining declaration and is independent of the other declarations of `val`.
+Entirely removing the declarations of `val` from `set_val` and `main` would result in a compile error that `val` is undeclared in `set_val` and `main`.
+
+If a variable referred to as `extern` is in the same file, it can have either internal or external linkage.
+Defining `val` as `static int val;` would have no effect on the use of `val` in `set_val` or `main`, except that the definition would need to be moved above them to compile.
+But if `val` were defined above the functions they would not need to declare `val` as `extern`.
+
+The following would work
+
+```c
+#include <stdio.h>
+
+// val defining declaration before the function definitions
+static int val;
+
+void set_val() {
+    val += 42;
+    // prints val is 42
+    printf("val is %d\n", val);
+}
+
+int main() {
+    set_val();
+    val += 42;
+    // prints val is 84
+    printf("val is %d\n", val);
+}
+```
+
+The `static` could be removed from `static int val;`, giving `val` external linkage,  and `val` would still work the same in `set_val` and `main`.
+If another file included this file, it could only use `val` if `val` had external linkage (not declared as `static`) and the other file declared `extern int val;`.
+
+A variable referred to by `extern` must not only have `static` storage, but must also have file scope.
+The following example will most likely not compile, because `val`, though `static`, does not have file scope.
+
+```c
+#include <stdio.h>
+
+void set_val() {
+    // defined with static storage, but not in file scope
+    static int val;
+    val += 42;
+    printf("val is %d\n", val);
+}
+
+int main() {
+    set_val();
+    extern int val;
+    printf("val is %d\n", val);
+}
+```
+
+## register (how to possibly speed access to a variable)
+
+A variable marked as `register` expresses the programmer's desire to have the value placed in a register for quick access.
+A `register` variable is like an `auto` variable in that it must be in function or block scope.
+Since the value is intended to be placed in a register instead of memory, accessing the address of the variable should be disallowed by the compiler, since the adress of a register can't be taken.
+However, a memory address itself can be placed in a register.
+The following example demonstrates that
+
+```c
+#include <stdio.h>
+
+int main() {
+    int i = 42;
+    register int *i_ptr = &i;
+    // prints i is 42, i_ptr is 0x7ffd0c2055c4 (or some other address)
+    printf("i is %d, i_ptr is %p", i, i_ptr);
+}
+```
+
+`register` is essentially a hint, since the value may or may not be actually placed in a register.
```

Comment:
This is compiler dependent.

You don't see `register` much used in modern code intended to run in a hosted environment (in an OS) because compilers that are intended for such environments are free to ignore  `register` if they like (they will usually be more efficient at deciding what should go in registers or not than humans). Other compilers my specifically always obey this specifier however (even if the humans are wrong 😄 ).

```suggestion
register` is essentially a hint, since compilers are free to chose if they follow this specifier or not, so the value may or may not be actually placed in a register.
```